Tools/test_shard_filter.py

```
@@ -0,0 +1,383 @@
+    tests = parse_tests(lines)
+
+    if not tests:
+        print("Error: no tests discovered from input", file=sys.stderr)
+        sys.exit(1)
+
+    class_counts = extract_classes(tests)
+    print(f"Discovered {len(tests)} tests in {len(class_counts)} classes, distributing across {total} shards", file=sys.stderr)
+
+    os.makedirs(output_dir, exist_ok=True)
+
+    def class_weight(cls):
```

Ruff (0.15.9)

[warning] Missing return type annotation for private function `class_weight`

(ANN202)